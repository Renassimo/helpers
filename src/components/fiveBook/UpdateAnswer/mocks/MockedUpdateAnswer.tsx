const MockedUpdateAnswer = jest.fn(() => <div>Mocked Update Answer:</div>);

export default MockedUpdateAnswer;
