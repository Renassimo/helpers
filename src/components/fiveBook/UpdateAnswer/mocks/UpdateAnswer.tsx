const MockedUpdateAnswer = jest.fn(() => <div>Mocked Change Answer:</div>);

export default MockedUpdateAnswer;
