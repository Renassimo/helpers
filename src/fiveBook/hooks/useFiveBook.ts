import { useContext } from 'react';

import FiveBookContext from '@/fiveBook/contexts/FiveBookContext';

const useFiveBook = () => useContext(FiveBookContext);

export default useFiveBook;
