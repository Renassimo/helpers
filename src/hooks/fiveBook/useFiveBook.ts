import { useContext } from 'react';

import FiveBookContext from '@/contexts/fiveBook';

const useFiveBook = () => useContext(FiveBookContext);

export default useFiveBook;
