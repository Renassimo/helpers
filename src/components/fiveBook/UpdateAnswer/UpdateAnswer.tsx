import { useCallback, useMemo, useRef } from 'react';
import { useRouter } from 'next/router';

import { TextField } from '@mui/material';
import Typography from '@mui/material/Typography';

import Modal from '@/components/common/Modal';

import useFiveBook from '@/hooks/fiveBook/useFiveBook';
import useUpdateAnswers from '@/hooks/fiveBook/useUpdateAnswers';

const UpdateAnswer = () => {
  const { answers, question } = useFiveBook();
  const { update, loading } = useUpdateAnswers(false);

  const { query, replace } = useRouter();
  const { updateYear } = query;

  const year = useMemo(
    () => (typeof updateYear == 'string' ? updateYear : null),
    [updateYear]
  );

  const answer = useMemo(
    () => answers?.find((answer) => answer.year === year)?.value ?? '',
    [answers, year]
  );

  const modalTitle = useMemo(
    () => (year ? `Update ${year} answer` : ''),
    [year]
  );

  const isModalOpen = useMemo(
    () => !!(year && answers?.find((answer) => answer.year === year)),
    [answers, year]
  );

  const handleModalClose = useCallback(async () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { updateYear, ...updatedQuery } = query;
    await replace({ pathname: '', query: updatedQuery }, undefined, {
      shallow: true,
    });
  }, [query, replace]);

  const updateAnswerRef = useRef<HTMLInputElement>(null);

  const onSubmit = useCallback(async () => {
    // @ts-ignore
    await update({ [String(year)]: updateAnswerRef?.current?.value });
    await handleModalClose();
  }, [handleModalClose, update, year]);

  return (
    <Modal
      open={isModalOpen}
      onClose={handleModalClose}
      title={modalTitle}
      onSubmit={onSubmit}
      loading={loading}
    >
      <Typography component="h3" variant="h6">
        {question}
      </Typography>
      <TextField
        id="updateAnswer"
        label="Update"
        inputRef={updateAnswerRef}
        multiline
        rows={4}
        fullWidth
        margin="dense"
        defaultValue={answer}
        autoFocus
        name="updateAnswer"
      />
    </Modal>
  );
};

export default UpdateAnswer;
