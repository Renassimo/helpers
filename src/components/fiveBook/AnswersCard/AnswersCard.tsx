import { MouseEvent, useCallback } from 'react';
import { useRouter } from 'next/router';

import Typography from '@mui/material/Typography';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import Link from '@mui/material/Link';

import FiveBookCard from '@/components/fiveBook/FiveBookCard';
import UpdateAnswer from '@/components/fiveBook/UpdateAnswer';

import useFiveBook from '@/hooks/fiveBook/useFiveBook';

const AnswersCard = () => {
  const { answers, question } = useFiveBook();
  const { replace, query } = useRouter();

  const handleModalOpen = useCallback(
    async (event: MouseEvent<HTMLElement>, year: string) => {
      event.preventDefault();
      await replace(
        { pathname: '', query: { ...query, updateYear: year } },
        undefined,
        { shallow: true }
      );
    },
    [query, replace]
  );

  return (
    <>
      <FiveBookCard>
        <Typography component="h2" variant="h6" textAlign="center" m={2}>
          {question}
        </Typography>
        <List>
          {answers.map(
            ({ year, value }) =>
              value && (
                <ListItem key={year} dense>
                  <ListItemText secondary={value}>
                    <Link
                      href={`?updateYear=${year}`}
                      onClick={(event) => handleModalOpen(event, year)}
                    >
                      {year}
                    </Link>
                  </ListItemText>
                </ListItem>
              )
          )}
        </List>
      </FiveBookCard>
      <UpdateAnswer />
    </>
  );
};

export default AnswersCard;
