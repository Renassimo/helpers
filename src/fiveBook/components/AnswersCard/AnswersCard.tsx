import { MouseEvent, useCallback } from 'react';
import { useRouter } from 'next/router';

import Typography from '@mui/material/Typography';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import Link from '@mui/material/Link';

import FiveBookCard from '@/fiveBook/components/FiveBookCard';
import UpdateAnswer from '@/fiveBook/components/UpdateAnswer';

import useFiveBook from '@/fiveBook/hooks/useFiveBook';

const AnswersCard = () => {
  const { answers, question } = useFiveBook();
  const { replace, query } = useRouter();

  const handleModalOpen = useCallback(
    async (event: MouseEvent<HTMLElement>, year: string) => {
      event.preventDefault();
      await replace(
        { pathname: '', query: { ...query, updateYear: year } },
        undefined,
        { shallow: true }
      );
    },
    [query, replace]
  );

  return (
    <>
      <FiveBookCard>
        <Typography component="h2" variant="h6" textAlign="center" m={2}>
          {question}
        </Typography>
        <List>
          {answers.map(
            ({ year, value }) =>
              value && (
                <ListItem key={year} dense>
                  <ListItemText secondary={value}>
                    <Link
                      href={`?updateYear=${year}`}
                      onClick={(event) => handleModalOpen(event, year)}
                    >
                      {year}
                    </Link>
                  </ListItemText>
                </ListItem>
              )
          )}
        </List>
      </FiveBookCard>
      <UpdateAnswer />
    </>
  );
};

export default AnswersCard;
