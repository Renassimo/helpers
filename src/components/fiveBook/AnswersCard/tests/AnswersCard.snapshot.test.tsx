import renderWithTheme from '@/tests/helpers';

import UpdateAnswer from '@/components/fiveBook/UpdateAnswer';
import AnswersCard from '@/components/fiveBook/AnswersCard';

import useFiveBook from '@/hooks/fiveBook/useFiveBook';

import MockedUpdateAnswer from '@/components/fiveBook/UpdateAnswer/mocks';

jest.mock('@/hooks/fiveBook/useFiveBook');
jest.mock('@/components/fiveBook/UpdateAnswer');
jest.mock('next/router', () => ({
  useRouter: () => ({
    replace: jest.fn(),
    query: { updateYear: '' },
  }),
}));

describe('AnswersCard snapshot', () => {
  const mockedQuestion = 'wSup?';
  const mockedAnswers = [
    {
      year: '2022',
      value: 'Good!',
    },
    {
      year: '2023',
      value: 'Nice!',
    },
  ];

  beforeEach(() => {
    const mockUseFiveBook = {
      answers: mockedAnswers,
      question: mockedQuestion,
    };
    (useFiveBook as jest.Mock).mockImplementation(() => mockUseFiveBook);
    (UpdateAnswer as jest.Mock).mockImplementation(() => MockedUpdateAnswer);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arrange
    // Act
    const { baseElement } = renderWithTheme(<AnswersCard />);
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(UpdateAnswer).toHaveBeenCalled();
  });
});
