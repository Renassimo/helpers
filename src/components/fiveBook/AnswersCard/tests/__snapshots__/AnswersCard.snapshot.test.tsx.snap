// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswersCard snapshot renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="css-w0nmsf"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 css-1fhdopv-MuiTypography-root"
        >
          wSup?
        </h2>
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
                  href="?updateYear=2022"
                >
                  2022
                </a>
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                Good!
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
                  href="?updateYear=2023"
                >
                  2023
                </a>
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                Nice!
              </p>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
